dimod/include/dimod/constraint.h

```diff
@@ -0,0 +1,178 @@
+//    limitations under the License.
+
+#pragma once
+
+#include <limits>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "dimod/abc.h"
+#include "dimod/expression.h"
+#include "dimod/vartypes.h"
+
+namespace dimod {
+
+enum Sense { LE, GE, EQ };
+
+enum Penalty { LINEAR, QUADRATIC, CONSTANT };
+
+template <class Bias, class Index>
+class Constraint : public Expression<Bias, Index> {
+ public:
+    /// The type of the base class.
+    using base_type = Expression<Bias, Index>;
+
+    /// The first template parameter (Bias).
+    using bias_type = Bias;
+
+    /// The second template parameter (Index).
+    using index_type = Index;
+
+    using size_type = typename base_type::size_type;
+
+    using parent_type = ConstrainedQuadraticModel<bias_type, index_type>;
+
+    Constraint();
+    explicit Constraint(parent_type* parent);
+
+    bool is_onehot() const;
+    bool is_soft() const;
+    void mark_discrete(bool mark = true);
+    bool marked_discrete() const;
+    Penalty penalty() const;
+    bias_type rhs() const;
+
+    // note: flips sign when negative
+    void scale(bias_type scalar);
+
+    Sense sense() const;
+    void set_penalty(Penalty penalty);
+    void set_rhs(bias_type rhs);
+    void set_sense(Sense sense);
+    void set_weight(bias_type weight);
+    bias_type weight() const;
+
+ private:
+    Sense sense_;
+    bias_type rhs_;
+    bias_type weight_;
+    Penalty penalty_;
+
+    // marker(s) - these ar not enforced by code
+    bool marked_discrete_ = false;
+};
+
+template <class bias_type, class index_type>
+Constraint<bias_type, index_type>::Constraint() : Constraint(nullptr) {}
+
+template <class bias_type, class index_type>
+Constraint<bias_type, index_type>::Constraint(parent_type* parent)
+        : base_type(parent),
+          sense_(Sense::EQ),
+          rhs_(0),
+          weight_(std::numeric_limits<bias_type>::infinity()),
+          penalty_(Penalty::LINEAR) {}
+
+template <class bias_type, class index_type>
+bool Constraint<bias_type, index_type>::is_onehot() const {
+    // must be linear and must have at least two variables
+    if (!base_type::is_linear() || base_type::num_variables() < 2) return false;
+
+    // all of out variables must be binary
+    for (const auto& v : base_type::variables()) {
+        if (base_type::vartype(v) != Vartype::BINARY) return false;
+    }
+
+    // we get a bit cute here and check the linear biases using the
+    // underlying model, thereby bypassing the dict lookup. This is not
+    // super future-proof but faster.
+    for (size_type i = 0; i < base_type::num_variables(); ++i) {
+        if (base_type::base_type::linear(i) != rhs_) return false;
+    }
+
+    return true;
+}
+
+template <class bias_type, class index_type>
+bool Constraint<bias_type, index_type>::is_soft() const {
+    // some sort of tolerance value?
+    return weight_ != std::numeric_limits<bias_type>::infinity();
```

Comment:
In cqm we have a default weight but a separate variable is_soft, depending on that we either use the wieght or not.

Reply:
And that means multiple sources of truth